### ISSUE TITLE
ST11 - False positive when using CTE with an array of nested STRUCTs

### ISSUE THREAD

**Issue by nathadfield** [sqlfluff/sqlfluff#6558]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

In BigQuery with queries that attempt to unnest STRUCTs from complex arrays, SQLFluff can generate false ST11 errors. 

### Expected Behaviour

In the example provided below, no error should be raised.

### Observed Behaviour

Here's the exact error message.

```
L:  16 | P:  11 | ST11 | Joined table 'UNNEST(ft.generic_array) AS g' not
                       | referenced elsewhere in query
                       | [structure.unused_join]
```

### How to reproduce

To reproduce, simply run a SQLFluff lint command on the following SQL.

```
WITH fact_table AS (
  SELECT
    1 AS id,
    ARRAY<STRUCT<
      nested_array ARRAY<STRUCT<generic_field STRING>>
    >>[
      STRUCT([STRUCT("Value1" AS generic_field), STRUCT("Value2" AS generic_field)] AS nested_array),
      STRUCT([STRUCT("Value3" AS generic_field)] AS nested_array)
    ] AS generic_array,
)

SELECT
  ft.id,
  n.generic_field,
FROM fact_table AS ft
LEFT JOIN UNNEST(ft.generic_array) AS g
LEFT JOIN UNNEST(g.nested_array) AS n
;
```

### Dialect

BigQuery

### Version

sqlfluff v3.3.0
Python v3.11.4

### Configuration

```
[tool.sqlfluff.core]
exclude_rules = "LT05,LT14,CV02,CV12,ST06,ST07,ST10,RF01,RF02,AM08,JJ01"
large_file_skip_byte_limit = 0
[tool.sqlfluff.indentation]
tab_space_size = 2
[tool.sqlfluff.rules.convention.terminator]
multiline_newline = true
require_final_semicolon = true
[tool.sqlfluff.rules.convention.select_trailing_comma]
select_clause_trailing_comma = "require"
[tool.sqlfluff.rules.convention.quoted_literals]
preferred_quoted_literal_style = "double_quotes"
[tool.sqlfluff.rules.capitalisation.keywords]
capitalisation_policy = "upper"
[tool.sqlfluff.rules.capitalisation.functions]
extended_capitalisation_policy = "upper"
[tool.sqlfluff.rules.capitalisation.types]
extended_capitalisation_policy = "upper"
[tool.sqlfluff.rules.capitalisation.literals]
capitalisation_policy = "upper"
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
